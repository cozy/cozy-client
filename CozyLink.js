"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.chain = exports.default = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var CozyLink = /*#__PURE__*/function () {
  function CozyLink(requestHandler) {
    (0, _classCallCheck2.default)(this, CozyLink);

    if (typeof requestHandler === 'function') {
      this.request = requestHandler;
    }
  }

  (0, _createClass2.default)(CozyLink, [{
    key: "request",
    value: function request(operation, result, forward) {
      throw new Error('request is not implemented');
    }
  }]);
  return CozyLink;
}();

exports.default = CozyLink;

var toLink = function toLink(handler) {
  return typeof handler === 'function' ? new CozyLink(handler) : handler;
};

var defaultLinkHandler = function defaultLinkHandler(operation, result) {
  if (result) return result;else if (operation.execute) return operation.execute();else throw new Error("No link could handle operation ".concat(JSON.stringify(operation)));
};

var chain = function chain(links) {
  return [].concat((0, _toConsumableArray2.default)(links), [defaultLinkHandler]).map(toLink).reduce(concat);
};

exports.chain = chain;

var concat = function concat(firstLink, nextLink) {
  return new CozyLink(function (operation, result, forward) {
    var nextForward = function nextForward(op, res) {
      return nextLink.request(op, res, forward);
    };

    return firstLink.request(operation, result, nextForward);
  });
};