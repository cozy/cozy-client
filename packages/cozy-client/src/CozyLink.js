export default class CozyLink {
  constructor(requestHandler) {
    if (typeof requestHandler === 'function') {
      this.request = requestHandler
    }
  }

  request(operation, result, forward) {
    throw new Error('request is not implemented')
  }
}

const toLink = handler =>
  typeof handler === 'function' ? new CozyLink(handler) : handler

const defaultLinkHandler = (operation, result) => {
  if (result) return result
  else
    throw new Error(
      `No link could handle operation ${JSON.stringify(operation)}`
    )
}

export const chain = links =>
  [...links, defaultLinkHandler].map(toLink).reduce(concat)

const concat = (firstLink, nextLink) => {
  return new CozyLink((operation, result, forward) => {
    const nextForward = (op, res) => {
      return nextLink.request(op, res, forward)
    }
    return firstLink.request(operation, result, nextForward)
  })
}
