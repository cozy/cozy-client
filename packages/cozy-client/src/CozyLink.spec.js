import CozyLink, { chain } from './CozyLink'

describe('CozyLink', () => {
  it('should be chainable', () => {
    const link = chain([
      new CozyLink((operation, result = '', forward) => {
        return forward(operation, result + 'foo')
      }),
      new CozyLink((operation, result, forward) => {
        return forward(operation, result + 'bar')
      }),
      (operation, result) => {
        return result + 'baz'
      }
    ])
    expect(link.request('dummyOperation')).toBe('foobarbaz')
  })

  describe('default last link', () => {
    it('should throw an error when called without result', () => {
      const link = chain([
        new CozyLink((operation, result = '', forward) => {
          return forward(operation)
        })
      ])
      expect(() =>
        link.request('dummyOperation')
      ).toThrowErrorMatchingSnapshot()
    })

    it('should return the final result if tehre is one', () => {
      const link = chain([
        new CozyLink((operation, result, forward) => {
          return forward(operation, 'foo')
        })
      ])

      expect(link.request('dummyOperation')).toBe('foo')
    })

    it('should call the custom execution function', () => {
      const link = chain([
        new CozyLink((operation, result = '', forward) => {
          return forward(operation)
        })
      ])
      const spyFn = jest.fn(() => 'bar')
      expect(
        link.request({
          type: 'dummyOperation',
          execute: spyFn
        })
      ).toBe('bar')
      expect(spyFn).toHaveBeenCalled()
    })
  })
})
