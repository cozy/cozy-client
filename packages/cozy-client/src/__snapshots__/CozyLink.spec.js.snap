// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CozyLink default last link should throw an error when called without result 1`] = `"No link could handle operation \\"dummyOperation\\""`;
